syntax = "proto3";

package descarteslabs.vektorius.v1;

import "descarteslabs/common/proto/ibis/ibis.proto";

// A request for the schema of a table.
message SchemaRequest {

  // The name of the table.
  string table_name = 1;
  // The database scope for the name.
  string database = 2;
}

// A response that contains a table schema.
message SchemaResponse {

  // The schema for a table, JSON-encoded.
  descarteslabs.ibis.Schema schema = 1;
  string table_name = 2;
}

// A query job.
message Job {

  // The id for the Job.
  string job_id = 1;

  // The status of the Job.
  enum STATUS {
    UNKNOWN = 0;
    PENDING = 1;
    RUNNING = 2;
    DONE = 3;
    CANCELLED = 4;
  }

  STATUS status = 2;
}

// Byte-slice of an Arrow RecordBatch.
message RecordBatchSlice {

  // The byte-slice of the RecordBatch.
  bytes data = 1;

  // Indicates if this is the last slice of a RecordBatch.
  bool final = 2;
}

// StreamingStatus error.
message Error {
  string message = 1;
}

// SearchResponse status.
message SearchStatus {

  // The Job associated with the search.
  // May not be available once streaming has begun.
  Job job = 1;

  // Query results are now available.
  bool results_available = 2;

  // Total number of result rows if results_available.
  uint64 total_rows = 3;

  // (Opaque) Continuation token for beginning/resuming streaming of the results.
  // If empty and results_available, all results have been processed.
  // This value is encrypted and signed to ensure integrity and authorization.
  string continuation_token = 4;
}

// A request that initiates a search.
message SearchRequest {

  // The query that the Job should run. JSON-encoded.
  descarteslabs.ibis.Query query = 1;
}

// A response that contains the status information for the search.
message SearchResponse {

  oneof block {
    // Search status for the search;
    SearchStatus status = 1;

    // Byte-slice of Arrow RecordBatch data.
    RecordBatchSlice slice = 2;
  }
}

// A request to get the status of a search Job.
message SearchStatusRequest {

  // Continuation token for the search.
  string continuation_token = 1;
}

// A response that contains the updated status.
message SearchStatusResponse {

  // The updated status information.
  SearchStatus status = 1;
}

// A request to get a page of a search.
message SearchResultsRequest {

  // Opaque continuation token containing information to resume retrieval.
  string continuation_token = 1;

  // Offset from the continuation_token location.
  uint64 offset = 2;
}

// A response for a SearchResultsRequest, streaming one or more blocks.
// First and last blocks are always status (unless a connection is broken).
message SearchResultsResponse {
  oneof block {

    // Current search status as of this point in the results stream.
    SearchStatus status = 1;

    // Byte-slice of Arrow RecordBatch data.
    RecordBatchSlice slice = 2;
  }
}

// The Vektorius service definition.
service Vektorius {

  // Get the schema for a table.
  rpc GetSchema(SchemaRequest) returns (SchemaResponse);

  // Initiate a search Job.
  rpc Search(SearchRequest) returns (stream SearchResponse);

  // Get the status of a search Job.
  rpc SearchStatus(SearchStatusRequest) returns (SearchStatusResponse);

  // Get the results of a successful search Job.
  rpc SearchResults(SearchResultsRequest) returns (stream SearchResultsResponse);
}
